### PR TITLE
JavaScript: Add support for Google Cloud Spanner.

### PR DESCRIPTION
**Pull request by xiemaisi**
A quick search didn't turn up any open-source projects that use it, so the model hasn't been tested on real-world code, only the unit tests.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll [hunk 1]
```diff
@@ -387,3 +387,105 @@ private module Sequelize {
...
+  class DatabaseRunCall extends SqlExecution {
+    DatabaseRunCall() {
+      exists (string run | run = "run" or run = "runPartitionedUpdate" or run = "runStream" |
+        this = database().getAMethodCall(run)
+      )
+    }
+  }
+
+  /**
+   * A call to `Trnsaction.run` or `Database.runStream`.
```
Comment (asger-semmle) [inline javascript/ql/src/semmle/javascript/frameworks/SQL.qll]: Trnsaction -> Transaction

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll [hunk 2]
```diff
@@ -387,3 +387,105 @@ private module Sequelize {
...
+
+  /**
+   * A call to a Spanner method that executes a SQL query.
+   */
+  abstract class SqlExecution extends DatabaseAccess, DataFlow::InvokeNode {
+    /**
+     * Gets the position of the query argument; default is zero, which can be overridden
+     * by subclasses.
+     */
+    int getQueryArgumentPosition() {
```
Comment (asger-semmle) [inline javascript/ql/src/semmle/javascript/frameworks/SQL.qll]: It's not actually overridden anywhere and the surrounding module is private. It looks like you opted for overriding `getAQueryArgument` instead.

Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/frameworks/SQL.qll]: Yes, this is a tiny bit of speculative generality.

**Comment by xiemaisi** [comment]
Typo amended.

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll [hunk 3]
```diff
@@ -387,3 +387,105 @@ private module Sequelize {
...
+    }
+
+    override DataFlow::Node getAQueryArgument() {
+      result = getArgument(getQueryArgumentPosition()) or
+      result = getOptionArgument(getQueryArgumentPosition(), "sql")
+    }
+  }
+
+  /**
+   * A call to `Database.run` or `Database.runStream`.
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/frameworks/SQL.qll]: Make this docstring more abstract? It looks exhaustive, but the predicate also supports `runPartitionedUpdate`...

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll [hunk 4]
```diff
@@ -387,3 +387,105 @@ private module Sequelize {
...
+  class DatabaseRunCall extends SqlExecution {
+    DatabaseRunCall() {
+      exists (string run | run = "run" or run = "runPartitionedUpdate" or run = "runStream" |
+        this = database().getAMethodCall(run)
+      )
+    }
+  }
+
+  /**
+   * A call to `Transaction.run` or `Database.runStream`.
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/frameworks/SQL.qll]: Make this docstring more abstract? It looks exhaustive, but the predicate also supports `runUpdate`...

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll (hunk 3 above)
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/frameworks/SQL.qll]: Well spotted; fixed (also below).

**Pull request merged**